### ISSUE TITLE
square brackets are deleted in a link

### ISSUE THREAD

**Issue by anhofmann** [zammad/zammad#2437]
### Infos:

* Used Zammad version: 2.8.0-1544012236.272eecf9.xenial
* Installation method (source, package, ..): package
* Operating system: 16.04.5 LTS (Xenial Xerus)
* Database + version: Postgresql 9.5+173ubuntu0.2
* Elasticsearch version: 5.6.13  
* Browser + version: 60.4.0esr (64-bit)

### Expected behavior:

* When clicking on a link in a ticket that looks like this: https://example.com/?foo=bar&baz[x]=y the link should be opened in the same way in a browser tab

### Actual behavior:

* When clicking a link in a ticket that looks like this: https://example.com/?foo=bar&baz[x]=y only the following part will be opened in a browser tab: https://example.com/?foo=bar&baz=y

### Steps to reproduce the behavior:

1. create a ticket in Zammad with the following link in the text: https://example.com/?foo=bar&baz[x]=y
2. open the ticket and click on the link.

[x] Yes I'm sure this is a bug and no feature request or a general question.

**Comment by anhofmann** [comment]
I looked around a bit in the code, but couldn't find where this error could occur. I'm not a Ruby developer either, so I don't know much about the code base. If someone could give me a hint in which file or directory the relevant code could be, I could try to create a PR.

**Comment by martinvonwittich** [comment]
I was curious and so I tested it myself, and I can confirm this issue. The bug is probably somewhere in the e-mail parser, because the brackets are already gone when I inspect the article in the database:

```
zammad_production=# select ta.body from tickets t join ticket_articles ta on (ta.ticket_id = t.id) where t.id = 7140;
-[ RECORD 1 ]----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
body | <div dir="ltr">                                                                                                                                                                                     +
     | <a href="https://example.com/?foo=bar&amp;baz=y" rel="nofollow noreferrer noopener" title="https://example.com/?foo=bar&amp;baz=y" target="_blank">https://example.com/?foo=bar&amp;baz[x]=y</a><br>+
     | </div>
```

**Comment by martinvonwittich** [comment]
This is my first foray into the Zammad source and I'm not very familiar with Ruby, so take everything I'm saying with a grain of salt.

I've looked through the source for files containing "parse", and I believe this is the email parser: https://github.com/zammad/zammad/blob/develop/app/models/channel/email_parser.rb

At one point, it calls a function named `html2html_strict`, for example:

```
  def body_text(message, **options)
    body_text = begin
                  message.body.to_s
                rescue Mail::UnknownEncodingType # see test/data/mail/mail043.box / issue #348
                  message.body.raw_source
                end

    body_text = body_text.utf8_encode(from: message.charset, fallback: :read_as_sanitized_binary)
    body_text = Mail::Utilities.to_lf(body_text)

    # plaintext body requires no processing
    return body_text if !options[:strict_html]

    # Issue #2390 - emails with >5k HTML links should be rejected
    return EXCESSIVE_LINKS_MSG if body_text.scan(/<a[[:space:]]/i).count >= 5_000

    body_text.html2html_strict
  end
```

This made me curious, because it's very probable that the braces are falling victim to some kind of HTML purifying process that aims to remove dangerous elements like JavaScript. `html2html_strict` is defined here: https://github.com/zammad/zammad/blob/develop/lib/core_ext/string.rb , but unfortunately it's complex and uses another class called `HtmlSanitizer`. At this point I just went ahead and attempted to use this function in the rails console, and there we have it:

```
irb(main):003:0> '<a href="https://example.com/?foo=bar&amp;baz[x]=y">https://example.com/?foo=bar&amp;baz[x]=y</a>'.html2html_strict
=> "<a href=\"https://example.com/?foo=bar&amp;baz=y\" rel=\"nofollow noreferrer noopener\" target=\"_blank\" title=\"https://example.com/?foo=bar&amp;baz=y\">https://example.com/?foo=bar&amp;baz[x]=y</a>"
```

So `html2html_strict` is definitely at fault. I went ahead and also called HtmlSanitizer directly, just like `html2html_strict` does:

```
irb(main):004:0> HtmlSanitizer.cleanup_replace_tags('<a href="https://example.com/?foo=bar&amp;baz[x]=y">https://example.com/?foo=bar&amp;baz[x]=y</a>')
=> "<a href=\"https://example.com/?foo=bar&amp;baz%5Bx%5D=y\">https://example.com/?foo=bar&amp;baz[x]=y</a>"
irb(main):005:0> HtmlSanitizer.strict('<a href="https://example.com/?foo=bar&amp;baz[x]=y">https://example.com/?foo=bar&amp;baz[x]=y</a>') 
=> "<a href=\"https://example.com/?foo=bar&amp;baz=y\" rel=\"nofollow noreferrer noopener\" target=\"_blank\" title=\"https://example.com/?foo=bar&amp;baz=y\">https://example.com/?foo=bar&amp;baz[x]=y</a>"
```

So `HtmlSanitizer.cleanup_replace_tags` is fine, but `HtmlSanitizer.strict` breaks it.

[`HtmlSanitizer`](https://github.com/zammad/zammad/blob/develop/lib/html_sanitizer.rb) is even more complex, but it definitely does something to `href`:

```
        # prepare links
        if node['href']
          href                = cleanup_target(node['href'], keep_spaces: true)
          href_without_spaces = href.gsub(/[[:space:]]/, '')
          if external && href_without_spaces.present? && !href_without_spaces.downcase.start_with?('//') && href_without_spaces.downcase !~ %r{^.{1,6}://.+?}
            node['href']        = "http://#{node['href']}"
            href                = node['href']
            href_without_spaces = href.gsub(/[[:space:]]/, '')
          end

          next if !href_without_spaces.downcase.start_with?('http', 'ftp', '//')

          node.set_attribute('href', href)
          node.set_attribute('rel', 'nofollow noreferrer noopener')
          node.set_attribute('target', '_blank')
        end
```

`cleanup_target` sounds interesting, what does it do?

```
  def self.cleanup_target(string, **options)
    cleaned_string = CGI.unescape(string).utf8_encode(fallback: :read_as_sanitized_binary)
    cleaned_string = cleaned_string.gsub(/[[:space:]]/, '') if !options[:keep_spaces]
    cleaned_string = cleaned_string.strip
                                   .delete("\t\n\r\u0000")
                                   .gsub(%r{/\*.*?\*/}, '')
                                   .gsub(/<!--.*?-->/, '')
                                   .gsub(/\[.+?\]/, '')

    sanitize_attachment_disposition(cleaned_string)
  end
```

Well, there's the bug. The fact that this occurs during parsing unfortunately implies that a lot of articles might be affected by this bug, and the only way to fix them is to re-parse the mails. I'm not sure if that's possible :(

**Comment by SummerStorm** [comment]
Hi, there. Square brackets should not appear in the latter part of the URI according to [RFC 3986](https://www.ietf.org/rfc/rfc3986.txt):

> A host identified by an Internet Protocol literal address, version 6 [RFC3513] or later, is distinguished by enclosing the IP literal within square brackets ("[" and "]"). This is the only place where square bracket characters are allowed in the URI syntax.

Which means that `<a href="https://example.com/?foo=bar&baz[x]=y">example</a>` is technically invalid HTML. 

In actual usage, Zammad users compose the HTML content in Email clients and word processors, and when you paste the HTML snippet `<a href="https://example.com/?foo=bar&baz[x]=y">example</a>` into any of these processors, they will smartly URL encode the offending URL, thus converting the snippet into valid HTML: `<a href="https://example.com/?foo=bar&baz%5Bx%5D=y">example</a>`. Zammad has no problem parsing the correctly encoded version.

So in conclusion I'm marking this as "won't fix" because the issue is unlikely to happen in actual usage:

1. Non-technical users will have their Email client/word processor automatically URL encode the offending URL 

2. Technical users can figure out it's due to malformed URL (likely thanks to this report) and then can fix the source that's generating the invalid HTML

**Comment by anhofmann** [comment]
Well, I wouldn't open a issue if I didn't have that problem all the time. We get mails from users who have such square brackets in the links without them being encoded. They just copy the URL and put it in their mail and that's how they arrive at Zammad.

**Comment by martinvonwittich** [comment]
> Non-technical users will have their Email client/word processor automatically URL encode the offending URL

I was able to reproduce this issue just by sending a mail with GMail, so it certainly didn't encode the brackets correctly :(
	
The source of the mail in GMail looks like this:

```
MIME-Version: 1.0
Date: Fri, 25 Jan 2019 21:16:12 +0100
Message-ID: <CAF49vrRKxJ5PMURYr7n9VB1kZ_e-f+wSTEcFpz5AgSSEBmyELg@mail.gmail.com>
Subject: test
From: Martin von Wittich <redacted@gmail.com>
To: Martin von Wittich via Zammad <zammad@ourcopmany.tld>
Content-Type: multipart/alternative; boundary="000000000000984ebf05804dff97"

--000000000000984ebf05804dff97
Content-Type: text/plain; charset="UTF-8"

https://example.com/?foo=bar&baz[x]=y

--000000000000984ebf05804dff97
Content-Type: text/html; charset="UTF-8"

<div dir="ltr"><a href="https://example.com/?foo=bar&amp;baz[x]=y">https://example.com/?foo=bar&amp;baz[x]=y</a><br></div>

--000000000000984ebf05804dff97--
```

Also, there's [RFC 1122](https://tools.ietf.org/html/rfc1122):

>                 "Be liberal in what you accept, and
>                 conservative in what you send"

:)

**Comment by SummerStorm** [comment]
Sorry, I spoke too soon. I can indeed re-produce this from Gmail. Do you have mail from any other clients that exhibit this same behavior? It would be good to have a list here. 

I only tried Mozilla Thunderbird and it worked fine there so I incorrectly assumed that other Email clients would be equally competent.

**Comment by martinvonwittich** [comment]
> I only tried Mozilla Thunderbird and it worked fine there so I incorrectly assumed that other Email clients would be equally competent.

Hmm, I think I can also reproduce the issue in Thunderbird? Not sure if I'm doing something wrong...? This is the source of the mail when I use Thunderbird with my gmail account and just send the mail to myself:

```
Return-Path: <redacted@gmail.com>
Received: from ?IPv6:2003:71:6f24:6600:redacted? (redacted.dip0.t-ipconnect.de. [2003:71:6f24:6600:redacted])
        by smtp.googlemail.com with ESMTPSA id n5sm107729415wrr.94.2019.01.26.02.19.36
        for <redacted@googlemail.com>
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Sat, 26 Jan 2019 02:19:36 -0800 (PST)
From: Martin von Wittich <redacted@gmail.com>
X-Google-Original-From: Martin von Wittich <redacted@googlemail.com>
To: redacted@googlemail.com
Subject: [] thunderbird
Message-ID: <423d0e29-1b83-5141-6c03-263d645d6ee5@googlemail.com>
Date: Sat, 26 Jan 2019 11:19:36 +0100
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Thunderbird/60.4.0
MIME-Version: 1.0
Content-Type: multipart/alternative; boundary="------------5BAC026C45528CC33D205880"
Content-Language: en-GB

--------------5BAC026C45528CC33D205880
Content-Type: text/plain; charset=utf-8; format=flowed
Content-Transfer-Encoding: 7bit

https://example.com/?foo=bar&baz[x]=y


--------------5BAC026C45528CC33D205880
Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: 7bit

<html>
  <head>

    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  </head>
  <body text="#000000" bgcolor="#FFFFFF">
    <p><a moz-do-not-send="true"
        href="https://example.com/?foo=bar&amp;baz[x]=y">https://example.com/?foo=bar&amp;baz[x]=y</a><br>
    </p>
  </body>
</html>

--------------5BAC026C45528CC33D205880--
```

When reproducing, you have to take care when copying the link. If you right-click on the link -> copy link address in anhofmann's OP, you'll have the link already URI encoded in your clipboard due to the GitHub's URI encoding. You have to copy it by selecting it -> copy so that it's not already pre-encoded, and then paste it into the MUA to see if the MUA encodes it or not. Maybe that's why you weren't able to reproduce it in TB?

**Comment by martinvonwittich** [comment]
The t-online.de webmailer also seems to be affected: [t-online.txt](https://github.com/zammad/zammad/files/2798899/t-online.txt)

**Comment by martinvonwittich** [comment]
web.de webmailer also affected: [web.de.txt](https://github.com/zammad/zammad/files/2798905/web.de.txt) (also no plaintext part, WTF, web.de?)

**Comment by martinvonwittich** [comment]
yahoo.de webmailer also affected: [yahoo.txt](https://github.com/zammad/zammad/files/2798922/yahoo.txt)
